---
title: "cell_lines"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{cell_lines}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(demuxSNPpaperfigures)
```



````{r}
library(demuxmix)
library(demuxSNP)
library(Matrix)
library(dittoSeq)
library(ggpubr)
library(gridExtra)
````


````{r}

source("../R/plot_hashtag.R")

````


````{r}

tenx_path<-c('/data/demuxSNP/data/processed/cellline_4x_totalseqanuc/outs/filtered_feature_bc_matrix')
#barcodes_path<-args[2]#c("/data/scratch/nextflow/ccrcc_out/barcodes_merged_e.tsv")
#lookup_path<-args[3]#c("/data/scratch/nextflow/ccrcc_out/lookup_table_doublets_pbmc_e_25pc.tsv")
#key<-args[4]#c("e")

matrix<-readMM(paste(tenx_path,"/matrix.mtx.gz",sep=""))
barcodes<-read.table(paste(tenx_path,"/barcodes.tsv.gz",sep=""))
features<-read.table(paste(tenx_path, "/features.tsv.gz",sep=""))

rownames(matrix)<-features$V2
colnames(matrix)<-barcodes$V1

rna<-matrix[features$V3=="Gene",]
hto<-matrix[grep("Hashtag",features$V1),]



sce<-SingleCellExperiment(list(counts=rna))
hto_sce<-SingleCellExperiment(list(counts=hto))
altExp(sce,"HTO")<-hto_sce
mainExpName(sce)<-"RNA"
head(colnames(sce))

````




````{r}

df<-data.frame(log(t(as.matrix(counts(altExp(sce,"HTO")))))+1)
plot<-plot_hashtag(df)
do.call(grid.arrange,c(plot,nrow=1))
````


````{r}
library(cellhashR)

totalseqa<-GenerateCellHashingCalls(as.matrix(hto),methods=c("htodemux",'bff_raw','bff_cluster'),doTSNE=FALSE)

````




````{r}

library(demuxSNP)
top_genes<-common_genes(sce = sce)
vcf_path<-c("/data/scratch/nextflow/data/genome1K.phase3.SNP_AF5e2.chr1toX.hg38.chr.vcf")
# subset
vcf<-readVcf(vcf_path,genome="GRCh38")
ensdb <- EnsDb.Hsapiens.v86::EnsDb.Hsapiens.v86
seqlevelsStyle(ensdb)<-"UCSC"
vcf_sub<-subset_vcf(vcf, top_genes, ensdb)

# print new vcf
writeVcf(vcf_sub,filename="vcf_sub.vcf")
getwd()
````



````{bash}

vartrix_linux --bam /data/demuxSNP/data/processed/cellline_4x_totalseqanuc/outs/possorted_genome_bam.bam --cell-barcodes /data/demuxSNP/data/processed/cellline_4x_totalseqanuc/outs/filtered_feature_bc_matrix/barcodes.tsv.gz --fasta /data/scratch/nextflow/refdata-gex-GRCh38-2020-A/fasta/genome.fa --vcf vcf_sub.vcf --threads 30

````




````{r}

mat<-readMM('out_matrix.mtx')
dim(mat)
mat<-as.matrix(mat)
thresh=0.5
mode(mat) <- "integer"
    mat_obs <- mat[(rowSums(mat > 0)/dim(mat)[2]) > thresh, ]
sce<-add_snps(sce,mat=mat,thresh=0.5)

library(ComplexHeatmap)
Heatmap(counts(altExp(sce,"SNP")),cluster_rows=FALSE)


````
