---
title: "Figure-4-application"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Figure-4-application}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(demuxSNPpaperfigures)
library(CiteFuse)
library(viridisLite)
library(dittoSeq)
library(SingleCellExperiment)
library(gridExtra)
```

````{r}

colors <- structure(viridis(n = 3), names = c("-1", "0", "1"))
colors_hashtags <- structure(dittoColors()[1:8], names = c("Hashtag1", "Hashtag2", "Hashtag3","Hashtag4","Hashtag5","Hashtag6","Doublet","Negative"))
````

````{r}

load("../data/sce_app.rda")

````


````{r}

library(scuttle)
df <- perCellQCMetrics(sce_app,
                       subsets=list(Mito=grep("^MT",rownames(sce_app))),
                       use_altexps=FALSE)
sce_app$lib_size<-df$sum
sce_app$percent_mito<-df$subsets_Mito_percent

````


````{r}
library(CiteFuse)
sce_app<-normaliseExprs(sce=sce_app,altExp_name = "HTO",transform="log")
sce_app<-crossSampleDoublets(sce_app)

library(Seurat)
seurat_app<-as.Seurat(sce_app,data=NULL)
seurat_app <- NormalizeData(seurat_app, assay = "HTO", normalization.method = "CLR")
seurat_app <- HTODemux(seurat_app, assay = "HTO", positive.quantile = 0.99)

table(seurat_app$hash.ID)
table(sce_app$doubletClassify_between_label,seurat_app$hash.ID)
#s2<-seurat_app[,seurat_app$nCount_RNA>1000 & seurat_app$percent_mito<10]
#s2<-HTODemux(s2, assay= "HTO")
#table(s2$hash.ID)
````



````{r}
df2<-as.data.frame(df[seurat_app$hash.ID=="Negative",])
df2$high_quality<-df2$sum>1500 & df2$detected>1000
library(ggpubr)
plots<-ggscatterhist(df2,x="sum",y="detected",color = "high_quality", title = "Quality metrics distributions for 'Negative' group")
plots$sp <- plots$sp +
    geom_hline(yintercept = 1000, linetype = "dashed", color = "blue") +
  geom_vline(xintercept = 1500, linetype = "dashed", color = "red") 

ggpar(plots,xscale="log10",yscale="log10",xlab=("library size"),ylab=c("detected genes"))
high_qual<-df2$sum>1500 & df2$detected>1000
ggpar(plots,xlab=("library size"),ylab=c("detected genes"))
high_qual<-df2$sum>1500 & df2$detected>1000
table(high_qual)

````



````{r fig.height=2,fig.width=7}

df_qual<-data.frame(table(high_qual))
df_qual$high_quality<-"high_quality"
ggbarplot(df_qual,x="high_quality",y="Freq", fill="high_qual",xlab=c(""),label = TRUE, lab.col = "black", lab.pos = "in",lab.vjust = c(0.5,0.5),lab.hjust = c(8,2),title = c("Numbers of high and low quality cells in HTODemux Negative group")) + rotate()+ylab("# cells")+xlab("")+rremove("y.text")
````



````{r}
sce<-sce_app
altExp(sce,"HTO")<-as(altExp(sce,"HTO"), "SingleCellExperiment")
library(Matrix)
library(demuxSNP)
mat<-as.matrix(readMM('../data/v_outs_full.mtx'))
sce<-add_snps(sce,mat=mat)
sce<-high_conf_calls(sce)
sce<-reassign(sce,k=20,d=50)
sce<-reassign_jaccard(sce,k=20,d=50)
sce<-reassign_balanced(sce,k=20)
#sce$reassign_jaccard<-reassign_centroid(sce,train_cells=sce$train,predict_cells = sce$predict)
sce$knn_centroid<-reassign_centroid(sce,predict_cells=!sce$train)
sce$hash.ID<-seurat_app$hash.ID
soup<-read.table(file="../data/clusters.tsv", sep='\t',header=TRUE)
soup$assignment[soup$status=="doublet"]<-"Doublet"
soup$assignment[soup$status=="unassigned"]<-"unassigned"
sce$soup<-soup$assignment

table(sce$knn_centroid)
table(sce$hash.ID)
table(sce$soup)

table(sce$soup,sce$knn_centroid)

sce$soup_rc<-recode(sce$soup,
                    "0"="Hashtag6",
                    "1"="Hashtag5",
                    "2"="Hashtag1",
                    "3"="Hashtag2",
                    "4"="Hashtag3",
                    "5"="Hashtag4")

table(sce$soup_rc,sce$knn_centroid)


````

# alluvial plot

````{r fig.width=12,fig.height=4}


library(ggalluvial)
df_all<-data.frame(soup=sce$soup_rc,demuxSNP=sce$knn_centroid,htodemux=sce$hash.ID)
df_allrs<-reshape2::melt(table(sce$soup_rc,sce$knn_centroid,sce$hash.ID))
colnames(df_allrs)<-c("souporcell","demuxSNP","HTODemux","ncells")
df_allrs$demuxSNP<-gsub('K','Hashtag',df_allrs$demuxSNP)
df_allrs$demuxSNP<-factor(df_allrs$demuxSNP,levels=c(paste("Hashtag",1:6,sep=""),"Doublet"))
ggplot(data = df_allrs,
       aes(axis1 = souporcell, axis2 = demuxSNP, axis3 = HTODemux, y = ncells)) +
  geom_alluvium(aes(fill = demuxSNP),alpha=0.9) +
  geom_stratum(alpha=1) +
  geom_text(stat = "stratum",
            aes(label = after_stat(stratum),angle=-20)) +
  scale_x_discrete(limits = c("Souporcell", "demuxSNP","HTODemux"),
                   expand = c(0.15, 0.05)) +
  scale_fill_manual(values=dittoColors()[1:8]) +
   coord_flip()

````


````{r}

table(sce$hash.ID)
table(sce$soup)
````


````{r}
library(demuxmix)
counts_low=as.matrix(counts(altExp(sce,"HTO")))
dmm<-demuxmix(hto=counts_low,model="naive")
pAcpt(dmm) <- 0
dmmlabels<-dmmClassify(dmm)
dmmlabels$HTO[dmmlabels$Type=="negative"]<-"Negative"
dmmlabels$HTO[dmmlabels$Type=="multiplet"]<-"Doublet"
split<-c(dmmlabels$Prob<0.9)

hto_jnorm<-NormalizeData(counts_low,normalization.method = "CLR")
df_j<-data.frame(t(hto_jnorm))
#plot_hashtag(df_j,split=split)

````
# Negative SNP heatmap

````{r}
library(ComplexHeatmap)
sce$knn_centroid<-gsub('K','Hashtag',sce$knn_centroid)
sce_negative<-sce[,sce$hash.ID=="Negative" & c(sce$percent_mito<15 | sce$lib_size>1000)]

ha = HeatmapAnnotation(demuxSNP = sce_negative$knn_centroid, 
                       HTODemux = sce_negative$hash.ID,
                       annotation_legend_param = list(
                          demuxSNP = list(
                              title = "Assignment"
                          )
                      ),
                      col=list(demuxSNP=colors_hashtags,HTODemux=colors_hashtags),
                      show_legend = c(FALSE,FALSE)
)
hm<-Heatmap(counts(altExp(sce_negative,"SNP")),
        cluster_rows = FALSE,
        column_split=sce_negative$knn_centroid,
        show_column_dend = FALSE,
        show_column_names = FALSE,
        cluster_column_slices = FALSE,
        column_title = "Cells",
        column_title_side = "bottom",
        row_title = "SNPs",
        show_row_names = FALSE,
        col = colors,
        top_annotation = ha,
        heatmap_legend_param = list(title="SNP status",at=c(-1,0,1),labels=c("Absent","No reads","Present")),
        )

draw(hm,
    column_title = "HTODemux high quality Negative SNP profiles split by demuxSNP call",
    padding = unit(c(2, 15, 2, 2), "mm")
)
````

# Negative distance matrix

````{r}

snps<-counts(altExp(sce,"SNP"))
snps[snps==c(-1)]<-0
library(geosphere)
centroid<-c()
combs1<-paste(sce$hash.ID, sce$knn_centroid[])
levels1<-grep("^Negative",unique(combs1),value=TRUE)
combs2<-paste(sce$train,sce$knn_centroid)
levels2<-grep("^TRUE",unique(combs2))
centroid1<-matrix(0,92,6)
for (i in c(1:6)) {
  hashtag<-hashtags[i]
  centroid1[,i]<-rowMeans(snps[,sce$hash.ID=="Negative" & sce$knn_centroid==hashtag])
}
colnames(centroid1)<-paste("Hashtag",c(1:6),"ds",sep="")
#centroid1
centroid2<-matrix(0,92,6)
for (i in c(1:6)) {
  hashtag<-hashtags[i]
  centroid2[,i]<-rowMeans(snps[,sce$train==TRUE & sce$knn_centroid==hashtag])
}
colnames(centroid2)<-paste("Hashtag", c(1:6),"hc",sep="")
#centroid2

c_full<-cbind(centroid1,centroid2)
c_fullnz<-c_full#[,colSums(c_full)!=0]

sim<-as.matrix(vegdist(t(c_fullnz),method="jaccard"),binary=TRUE)
dim(sim)
mat<-sim[grep("ds",rownames(sim)),grep("hc",rownames(sim))]
mat<-1-mat

````

````{r fig.height=5.5,fig.width=6}

col_fun<-colorRamp2(c(min(mat,na.rm = TRUE),max(mat,na.rm=TRUE)),c("white","brown"))
hm<-Heatmap(mat,
        cluster_rows = FALSE,
        cluster_columns = FALSE,
        col=col_fun,column_title = "High confidence cells",
        row_title = "Negative reassigned cells",
        heatmap_legend_param = list(title = "Jaccard Similarity",direction = "horizontal"),
        cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
        grid.text(round(mat[i, j],2), x, y)},
        column_names_rot = 0,
        column_names_centered = TRUE,
        show_heatmap_legend = FALSE,
        row_labels = substr(rownames(mat),1,8),
        column_labels = substr(colnames(mat),1,8)
    )
draw(hm,heatmap_legend_side = "bottom",column_title="Reassigned HTODemux Negative")
````


# Hashtag2 SNP heatmap

````{r}

sce_hashtag2<-sce[,sce$hash.ID=="Hashtag2"]
colors_hashtags <- structure(dittoColors()[1:8], names = c("Hashtag1", "Hashtag2", "Hashtag3","Hashtag4","Hashtag5","Hashtag6","Doublet","Negative"))

ha = HeatmapAnnotation(demuxSNP = sce_hashtag2$knn_centroid, 
                       HTODemux = sce_hashtag2$hash.ID,
                       annotation_legend_param = list(
                          demuxSNP = list(
                              title = "Assignment"
                          )
                      ),
                      col=list(demuxSNP=colors_hashtags,HTODemux=colors_hashtags),
                      show_legend = c(FALSE,FALSE)
)

hm<-Heatmap(counts(altExp(sce_hashtag2,"SNP")),
        cluster_rows = FALSE,
        column_split=sce_hashtag2$knn_centroid,
        show_column_dend = FALSE,
        show_column_names = FALSE,
        cluster_column_slices = FALSE,
        row_title = "SNPs",
        column_title="Cells",
        column_title_side = "bottom",
        show_row_names = FALSE,
        col = colors,
        heatmap_legend_param = list(title="SNP status",at=c(-1,0,1),labels=c("Absent","No reads","Present")),
        top_annotation = ha
        )


draw(hm,
    column_title = "HTODemux high quality Hashtag2 SNP profiles split by demuxSNP call",
    padding = unit(c(2, 15, 2, 2), "mm"),
    heatmap_legend_side="right",
    annotation_legend_side="right"
)
````

# Hashtag 2 binary distance hist and matrix

````{r}

snps<-counts(altExp(sce,"SNP"))
snps[snps==c(-1)]<-0
library(geosphere)
centroid<-c()
combs1<-paste(sce$hash.ID, sce$knn_centroid[])
levels1<-grep("^Hashtag2",unique(combs1),value=TRUE)
combs2<-paste(sce$train,sce$knn_centroid)
levels2<-grep("^TRUE",unique(combs2))
centroid1<-matrix(0,92,6)
for (i in c(2,3,4,5)) {
  hashtag<-hashtags[i]
  centroid1[,i]<-rowMeans(snps[,sce$hash.ID=="Hashtag2" & sce$knn_centroid==hashtag])
}
colnames(centroid1)<-paste("Hashtag",c(1:6),"ds",sep="")
#centroid1
centroid2<-matrix(0,92,6)
for (i in c(1:6)) {
  hashtag<-hashtags[i]
  centroid2[,i]<-rowMeans(snps[,sce$train==TRUE & sce$knn_centroid==hashtag])
}
colnames(centroid2)<-paste("Hashtag", c(1:6),"hc",sep="")
#centroid2

c_full<-cbind(centroid1,centroid2)
c_fullnz<-c_full#[,colSums(c_full)!=0]

sim<-as.matrix(vegdist(t(c_fullnz),method="jaccard"),binary=TRUE)
dim(sim)
mat<-sim[grep("ds",rownames(sim)),grep("hc",rownames(sim))]
mat<-1-mat
mat[c(1,6),]<-NA
````

````{r fig.height=5.5,fig.width=6}


col_fun<-colorRamp2(c(min(mat,na.rm=TRUE),max(mat,na.rm=TRUE)),c("white","brown"))
hm<-Heatmap(mat,
        cluster_rows = FALSE,
        cluster_columns = FALSE,
        col=col_fun,column_title = "High confidence cells",
        row_title = "Hashtag2 reassigned cells",
        heatmap_legend_param = list(title = "Euclidean distance",direction = "horizontal"),
        cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
        grid.text(round(mat[i, j],2), x, y)},
        column_names_rot = 0,
        column_names_centered = TRUE,
        show_heatmap_legend = FALSE,
        row_labels = substr(rownames(mat),1,8),
        column_labels = substr(colnames(mat),1,8)
    )
draw(hm,heatmap_legend_side = "bottom",column_title="Reassigned HTODemux Hashtag2")
````


````{r}
library(vegan)

hashtags<-paste("Hashtag",1:6,sep="")
plots<-list()
for (i in seq_along(hashtags)) {
ht<-hashtags[i]
snps<-t(counts(altExp(sce,"SNP")[,sce$hash.ID==ht & sce$lib_size>1500]))
snps_bin<-snps
snps_bin[snps==c(-1)]<-0

vdist<-vegdist(snps_bin,method="jaccard",binary=TRUE)
#hist(vdist)

dbm<-data.frame(binary_distance=matrix(vdist))
#plot(Heatmap(t(snps_bin),cluster_rows = FALSE,show_column_names = FALSE,show_row_names = FALSE))
gg<-gghistogram(dbm,x='binary_distance',xlab=ht,fill=dittoColors()[i],alpha = 1)
#plot(gg)
#gg2<-ggpar(gg, tickslab = FALSE)
gg2<-gg+rremove('y.text')+font('xlab',size=24)

plot(gg2)
  plots[[i]]<-gg2
}
````

````{r fig.height=7,fig.width=6}

grid.arrange(grobs=plots,nrow=3,top=textGrob("Within-group SNP Jaccard distance\ndistributions for HTODemux", gp=gpar(fontsize=18,font=8)))

````

````{r}
hashtags<-c("Hashtag1","Hashtag2","Hashtag3","Hashtag4","Hashtag5","Hashtag6")
for (i in seq_along(hashtags)) {
  ht<-hashtags[i]
  print(ht)
  match<-sum((as.character(sce$hash.ID)[sce$hash.ID==ht]==as.character(sce$knn_centroid)[sce$hash.ID==ht])/length(sce$hash.ID[sce$hash.ID==ht]))
  print(match)
}

````





````{r}

lgd = Legend(at = c("Hashtag1","Hashtag2","Hashtag3","Hashtag4","Hashtag5","Hashtag6","Doublet","Negative"), title = "Assignment", legend_gp = gpar(fill = dittoColors()[1:8]),nrow=2)
draw(lgd,test="")

col_fun<-colorRamp2(c(0,1),c("white","brown"))
lgd = Legend(col_fun = col_fun, title = "1-Jaccard distance", at = c(0, 1),
    labels = c("Dissimilar", "Similar"),
    direction = "horizontal")
draw(lgd,test="")
````


